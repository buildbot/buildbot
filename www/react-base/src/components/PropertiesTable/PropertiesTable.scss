
.bb-properties-value {
  vertical-align: top;
  padding: unset;
  margin: unset;
  display: inline-block;
  border: unset;
  background-color: unset;
}

.bb-properties-copy {
  vertical-align: top;
  padding: 0 0.5ex;
}
