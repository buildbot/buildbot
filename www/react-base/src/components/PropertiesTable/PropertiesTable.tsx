/*
  This file is part of Buildbot.  Buildbot is free software: you can
  redistribute it and/or modify it under the terms of the GNU General Public
  License as published by the Free Software Foundation, version 2.

  This program is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
  details.

  You should have received a copy of the GNU General Public License along with
  this program; if not, write to the Free Software Foundation, Inc., 51
  Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

  Copyright Buildbot Team Members
*/

import './PropertiesTable.scss';
import {Table} from "react-bootstrap";
import CopyToClipboard from 'react-copy-to-clipboard';
import {observer} from "mobx-react";
import {FaCopy} from "react-icons/fa";

type PropertiesTableProps = {
  properties: Map<string, any>;
}

export const PropertiesTable = observer(({properties}: PropertiesTableProps) => {
  const propertyRows = Array.from(properties.entries()).map(([key, valueSource]: [string, any]) => {
    const [value, source] = valueSource;
    const valueString = JSON.stringify(value);
    return (
      <tr key={key}>
        <td className="text-left">{key}</td>
        <td className="text-left">
          <pre className="bb-properties-value">{valueString}</pre>
          {/* @ts-ignore CopyToClipboard is not understood as React component for some reason */}
          <CopyToClipboard text={valueString}>
            <FaCopy className="bb-properties-copy clickable"></FaCopy>
          </CopyToClipboard>
        </td>
        <td className="text-right">{source}</td>
      </tr>
    )
  });

  return (
    <Table hover striped size="sm">
      <thead>
        <tr>
          <th className="text-left">Name</th>
          <th className="text-center">Value</th>
          <th className="text-right">Source</th>
        </tr>
      </thead>
      <tbody>
        {propertyRows}
      </tbody>
    </Table>
  );
});
