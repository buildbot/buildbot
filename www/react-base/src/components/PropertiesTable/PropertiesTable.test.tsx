/*
  This file is part of Buildbot.  Buildbot is free software: you can
  redistribute it and/or modify it under the terms of the GNU General Public
  License as published by the Free Software Foundation, version 2.

  This program is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
  details.

  You should have received a copy of the GNU General Public License along with
  this program; if not, write to the Free Software Foundation, Inc., 51
  Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

  Copyright Buildbot Team Members
*/

import renderer from 'react-test-renderer';
import {MemoryRouter} from "react-router-dom";
import {PropertiesTable} from './PropertiesTable';

function assertRenderSnapshot(properties: Map<string, any>) {
  const component = renderer.create(
    <MemoryRouter>
      <PropertiesTable properties={properties}/>
    </MemoryRouter>
  );
  expect(component.toJSON()).toMatchSnapshot();
}

describe('PropertiesTable component', function() {
  it('rendering', () => {
    const obj = {
      str: ["string", "string source"],
      int: [123, "int source"],
      float: [123.4, "float source"],
      boolean: [true, "boolean source"],
      array: [[321, "astr", false], "array source"],
      object: [{s: "str", i: 789, b: false}, "object source"],
    }

    assertRenderSnapshot(new Map(Object.entries(obj)));
  });
});
