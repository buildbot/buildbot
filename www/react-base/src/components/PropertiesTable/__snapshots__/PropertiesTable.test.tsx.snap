// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertiesTable component rendering 1`] = `
<table
  className="table table-sm table-striped table-hover"
>
  <thead>
    <tr>
      <th
        className="text-left"
      >
        Name
      </th>
      <th
        className="text-center"
      >
        Value
      </th>
      <th
        className="text-right"
      >
        Source
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        className="text-left"
      >
        str
      </td>
      <td
        className="text-left"
      >
        <pre
          className="bb-properties-value"
        >
          "string"
        </pre>
        <svg
          className="bb-properties-copy clickable"
          fill="currentColor"
          height="1em"
          onClick={[Function]}
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M320 448v40c0 13.255-10.745 24-24 24H24c-13.255 0-24-10.745-24-24V120c0-13.255 10.745-24 24-24h72v296c0 30.879 25.121 56 56 56h168zm0-344V0H152c-13.255 0-24 10.745-24 24v368c0 13.255 10.745 24 24 24h272c13.255 0 24-10.745 24-24V128H344c-13.2 0-24-10.8-24-24zm120.971-31.029L375.029 7.029A24 24 0 0 0 358.059 0H352v96h96v-6.059a24 24 0 0 0-7.029-16.97z"
          />
        </svg>
      </td>
      <td
        className="text-right"
      >
        string source
      </td>
    </tr>
    <tr>
      <td
        className="text-left"
      >
        int
      </td>
      <td
        className="text-left"
      >
        <pre
          className="bb-properties-value"
        >
          123
        </pre>
        <svg
          className="bb-properties-copy clickable"
          fill="currentColor"
          height="1em"
          onClick={[Function]}
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M320 448v40c0 13.255-10.745 24-24 24H24c-13.255 0-24-10.745-24-24V120c0-13.255 10.745-24 24-24h72v296c0 30.879 25.121 56 56 56h168zm0-344V0H152c-13.255 0-24 10.745-24 24v368c0 13.255 10.745 24 24 24h272c13.255 0 24-10.745 24-24V128H344c-13.2 0-24-10.8-24-24zm120.971-31.029L375.029 7.029A24 24 0 0 0 358.059 0H352v96h96v-6.059a24 24 0 0 0-7.029-16.97z"
          />
        </svg>
      </td>
      <td
        className="text-right"
      >
        int source
      </td>
    </tr>
    <tr>
      <td
        className="text-left"
      >
        float
      </td>
      <td
        className="text-left"
      >
        <pre
          className="bb-properties-value"
        >
          123.4
        </pre>
        <svg
          className="bb-properties-copy clickable"
          fill="currentColor"
          height="1em"
          onClick={[Function]}
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M320 448v40c0 13.255-10.745 24-24 24H24c-13.255 0-24-10.745-24-24V120c0-13.255 10.745-24 24-24h72v296c0 30.879 25.121 56 56 56h168zm0-344V0H152c-13.255 0-24 10.745-24 24v368c0 13.255 10.745 24 24 24h272c13.255 0 24-10.745 24-24V128H344c-13.2 0-24-10.8-24-24zm120.971-31.029L375.029 7.029A24 24 0 0 0 358.059 0H352v96h96v-6.059a24 24 0 0 0-7.029-16.97z"
          />
        </svg>
      </td>
      <td
        className="text-right"
      >
        float source
      </td>
    </tr>
    <tr>
      <td
        className="text-left"
      >
        boolean
      </td>
      <td
        className="text-left"
      >
        <pre
          className="bb-properties-value"
        >
          true
        </pre>
        <svg
          className="bb-properties-copy clickable"
          fill="currentColor"
          height="1em"
          onClick={[Function]}
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M320 448v40c0 13.255-10.745 24-24 24H24c-13.255 0-24-10.745-24-24V120c0-13.255 10.745-24 24-24h72v296c0 30.879 25.121 56 56 56h168zm0-344V0H152c-13.255 0-24 10.745-24 24v368c0 13.255 10.745 24 24 24h272c13.255 0 24-10.745 24-24V128H344c-13.2 0-24-10.8-24-24zm120.971-31.029L375.029 7.029A24 24 0 0 0 358.059 0H352v96h96v-6.059a24 24 0 0 0-7.029-16.97z"
          />
        </svg>
      </td>
      <td
        className="text-right"
      >
        boolean source
      </td>
    </tr>
    <tr>
      <td
        className="text-left"
      >
        array
      </td>
      <td
        className="text-left"
      >
        <pre
          className="bb-properties-value"
        >
          [321,"astr",false]
        </pre>
        <svg
          className="bb-properties-copy clickable"
          fill="currentColor"
          height="1em"
          onClick={[Function]}
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M320 448v40c0 13.255-10.745 24-24 24H24c-13.255 0-24-10.745-24-24V120c0-13.255 10.745-24 24-24h72v296c0 30.879 25.121 56 56 56h168zm0-344V0H152c-13.255 0-24 10.745-24 24v368c0 13.255 10.745 24 24 24h272c13.255 0 24-10.745 24-24V128H344c-13.2 0-24-10.8-24-24zm120.971-31.029L375.029 7.029A24 24 0 0 0 358.059 0H352v96h96v-6.059a24 24 0 0 0-7.029-16.97z"
          />
        </svg>
      </td>
      <td
        className="text-right"
      >
        array source
      </td>
    </tr>
    <tr>
      <td
        className="text-left"
      >
        object
      </td>
      <td
        className="text-left"
      >
        <pre
          className="bb-properties-value"
        >
          {"s":"str","i":789,"b":false}
        </pre>
        <svg
          className="bb-properties-copy clickable"
          fill="currentColor"
          height="1em"
          onClick={[Function]}
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M320 448v40c0 13.255-10.745 24-24 24H24c-13.255 0-24-10.745-24-24V120c0-13.255 10.745-24 24-24h72v296c0 30.879 25.121 56 56 56h168zm0-344V0H152c-13.255 0-24 10.745-24 24v368c0 13.255 10.745 24 24 24h272c13.255 0 24-10.745 24-24V128H344c-13.2 0-24-10.8-24-24zm120.971-31.029L375.029 7.029A24 24 0 0 0 358.059 0H352v96h96v-6.059a24 24 0 0 0-7.029-16.97z"
          />
        </svg>
      </td>
      <td
        className="text-right"
      >
        object source
      </td>
    </tr>
  </tbody>
</table>
`;
